### PR TITLE
feat: INTMDB-1103:  project-invitation

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes

_Jira ticket:_ [INTMDB-1104](https://jira.mongodb.org/browse/INTMDB-1104)

Migrating the `project_invitation` resource to use the new Atlas SDK.

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] This change requires a documentation update

## Manual QA performed:

- [x] cfn invoke for each of CRUDL/cfn test
- [x] Updated resource in  [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples)
- [x] Published to AWS private registry
- [x] Used the template in [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples) to create and update a stack in AWS
- [x] Deleted stack to ensure resources are deleted
- [x] Created multiple resources in same stack
- [x] Validated in Atlas UI
- [x] Included screenshots

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code
- [x] For CFN Resources: I have released by changes in the private registry and proved by change
  works in Atlas

#### CFN test
```bash
cfn test

============== 8 passed, 4 skipped, 12 deselected in 48.66s ==============
```

### CFN Stack
Screenshot 2023-09-25 at 12 18 21

Screenshot 2023-09-25 at 12 16 59

### CODE REVIEW

File: cfn-resources/project-invitation/cmd/resource/create.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util"
+	log "github.com/mongodb/mongodbatlas-cloudformation-resources/util/logger"
+	progressevents "github.com/mongodb/mongodbatlas-cloudformation-resources/util/progressevent"
+	"go.mongodb.org/atlas-sdk/v20230201008/admin"
+)
+
+func CreateOp(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup()
+
+	_, _ = log.Debugf("Create() currentModel:%+v", currentModel)
```
Comment (marcosuma) [inline cfn-resources/project-invitation/cmd/resource/create.go:32]: nit: no need for `_, _` I guess?

Comment (marcosuma) [reply cfn-resources/project-invitation/cmd/resource/create.go:32]: many other places that do that below

Comment (marcosuma) [reply cfn-resources/project-invitation/cmd/resource/create.go:32]: _(I know you didn't write this and moved it from other places)_

Comment (andreaangiolillo) [reply cfn-resources/project-invitation/cmd/resource/create.go:32]: > nit: no need for _, _ I guess?

i think we need it to make clear that the function is returning values but we deliberately ignore them

Comment (andreaangiolillo) [reply cfn-resources/project-invitation/cmd/resource/create.go:32]: I will update the others as well to make it consistent

Comment (andreaangiolillo) [reply cfn-resources/project-invitation/cmd/resource/create.go:32]: FYI: I will open a follow-up PR to update the other files with this format if that is okay with you. (there are a lot of files 😅 )

File: cfn-resources/project-invitation/test/inputs_1_invalid.template.json [hunk 2]
```diff
@@ -1,8 +1 @@
-{
```
Comment (maastha) [inline cfn-resources/project-invitation/test/inputs_1_invalid.template.json:1]: delete file?

Comment (andreaangiolillo) [reply cfn-resources/project-invitation/test/inputs_1_invalid.template.json:1]: we don't use invalid template in the cfn tests, we should delete all of them

File: cfn-resources/project-invitation/cmd/resource/create.go (hunk 1 above)
Comment (lantoli) [reply cfn-resources/project-invitation/cmd/resource/create.go:32]: I think _, _ is currently needed because of our linter rules

File: cfn-resources/project-invitation/cmd/resource/create.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+	modelValidation := validateModel(CreateRequiredFields, currentModel)
+	if modelValidation != nil {
```
Comment (lantoli) [inline cfn-resources/project-invitation/cmd/resource/create.go:35]: knitpit: i'd use err or similar for naming the err variable, at first I though modelValidation would have the model validated or something like that.

File: cfn-resources/project-invitation/cmd/resource/create.go (hunk 1 above)
Comment (andreaangiolillo) [reply cfn-resources/project-invitation/cmd/resource/create.go:32]: I see `log.Printf` that do not have it in the repo 🤔

File: cfn-resources/project-invitation/cmd/resource/create.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	invitationReq := &admin.GroupInvitationRequest{
+		Roles:    currentModel.Roles,
+		Username: currentModel.Username,
+	}
+
+	invitation, res, err := client.AtlasV2.ProjectsApi.CreateProjectInvitation(context.Background(), *currentModel.ProjectId, invitationReq).Execute()
+	if err != nil {
+		_, _ = log.Warnf("Create - error: %+v", err)
```
Comment (lantoli) [inline cfn-resources/project-invitation/cmd/resource/create.go:55]: not sure we want to log this if it'll be logged because it's an error

File: cfn-resources/project-invitation/cmd/resource/resource.go [hunk 5]
```diff
@@ -34,237 +30,59 @@ var UpdateRequiredFields = []string{constants.ProjectID, constants.ID}
...
-		return progressevents.GetFailedEventByResponse(err.Error(), res.Response), nil
-	}
-	_, _ = log.Debugf("%s invitation updated", *currentModel.Id)
-
-	// Response
-	return handler.ProgressEvent{
-		OperationStatus: handler.Success,
-		ResourceModel:   invitationToModel(currentModel, invitation),
-	}, nil
+	return UpdateOp(req, prevModel, currentModel)
```
Comment (lantoli) [inline cfn-resources/project-invitation/cmd/resource/resource.go:42]: one small issue with this refactor extracting methods is that it's hard to follow the changes as they're moved to other files

File: cfn-resources/project-invitation/cmd/resource/resource.go [hunk 6]
```diff
@@ -34,237 +30,59 @@ var UpdateRequiredFields = []string{constants.ProjectID, constants.ID}
...
 		ProjectId:       currentModel.ProjectId,
-		Username:        &invitation.Username,
-		Id:              &invitation.ID,
+		Username:        invitation.Username,
+		Id:              invitation.Id,
 		Roles:           invitation.Roles,
-		ExpiresAt:       &invitation.ExpiresAt,
-		CreatedAt:       &invitation.CreatedAt,
-		InviterUsername: &invitation.InviterUsername,
+		ExpiresAt:       util.StringPointerIfNotEmpty(invitation.ExpiresAt.GoString()),
```
Comment (lantoli) [inline cfn-resources/project-invitation/cmd/resource/resource.go:82]: why GoString and StringPointerIfNotEmpty?

**Review by lantoli — APPROVED** [review]

File: cfn-resources/project-invitation/cmd/resource/resource.go (hunk 6 above)
Comment (andreaangiolillo) [reply cfn-resources/project-invitation/cmd/resource/resource.go:82]: `invitation.ExpiresAt` is of type `*time.Time` but the model wants a `*string`

File: cfn-resources/project-invitation/cmd/resource/create.go (hunk 3 above)
Comment (andreaangiolillo) [reply cfn-resources/project-invitation/cmd/resource/create.go:35]: it makes sense. Updated it

File: cfn-resources/project-invitation/cmd/resource/create.go (hunk 4 above)
Comment (andreaangiolillo) [reply cfn-resources/project-invitation/cmd/resource/create.go]: yes, you're right. I removed it

File: cfn-resources/project-invitation/cmd/resource/resource.go (hunk 5 above)
Comment (lantoli) [reply cfn-resources/project-invitation/cmd/resource/resource.go:42]: I mean you can't compare easily old and new code

Comment (andreaangiolillo) [reply cfn-resources/project-invitation/cmd/resource/resource.go:42]: yes, this is expected. All the follow-up PRs related to this resource won't have the same issue

File: cfn-resources/project-invitation/cmd/resource/resource.go (hunk 6 above)
Comment (lantoli) [reply cfn-resources/project-invitation/cmd/resource/resource.go:82]: but I'm not sure that we're using ISO here, I think we should depend on util.TimeToString here, validating first if time ptr is not nil

Comment (andreaangiolillo) [reply cfn-resources/project-invitation/cmd/resource/resource.go]: Thanks! I didn't see TimeToString. Updated

File: cfn-resources/project-invitation/test/inputs_1_invalid.template.json (hunk 2 above)
Comment (maastha) [reply cfn-resources/project-invitation/test/inputs_1_invalid.template.json:1]: right, so we should delete this one in this PR right? currently this is an empty file

Comment (andreaangiolillo) [reply cfn-resources/project-invitation/test/inputs_1_invalid.template.json:1]: oh 🤦 thanks.  I will remove it xD

Comment (andreaangiolillo) [reply cfn-resources/project-invitation/test/inputs_1_invalid.template.json:1]: updated

Comment (AgustinBettati) [reply cfn-resources/project-invitation/test/inputs_1_invalid.template.json:1]: Small question:

> we don't use invalid template in the cfn tests

Is this something particular to how cfn runs contract tests, or a decision that was made from our side?

**Review by AgustinBettati — APPROVED** [review]

File: cfn-resources/project-invitation/test/inputs_1_invalid.template.json (hunk 2 above)
Comment (andreaangiolillo) [reply cfn-resources/project-invitation/test/inputs_1_invalid.template.json:1]: The invalid template was added initially by contractors but then they figured out that was not used by the test scenarios run as part of cfn test https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/contract-tests.html
